### PR TITLE
Playbooks for copy.sh and setup.sh

### LINKED ISSUE #294: ansible playbook for setup.sh
We need a playbook for replacing [setup.sh](https://github.com/AutolabJS/AutolabJS/blob/dev/deploy/setup.sh). The setup.sh script uses [setup.conf](https://github.com/AutolabJS/AutolabJS/blob/dev/deploy/setup.conf) for reading environment variables. The environment variables are to be read from inventory files, namely [single_machine](https://github.com/AutolabJS/AutolabJS/blob/dev/deploy/single_machine) and [two_machines](https://github.com/AutolabJS/AutolabJS/blob/dev/deploy/two_machines).

@shivin7 you can attempt this issue.

### PR DESCRIPTION
**Pull request by shivin7**
## Description
Creates a playbook `setup.yml` for commands in `setup.sh`

## Motivation
Fixes #294 

 ## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checks
 - [ ] Squash your commits down to one, or a few, discreet changesets before submitting a pull request
 - [ ] Read the [contributing guide](https://github.com/AutolabJS/AutolabJS/blob/master/CONTRIBUTING.md)
 - [ ] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [ ] If possible, include an example in the description, that could help all readers (including project members) to get a better picture of the change
 - [ ] Avoid other runtime dependencies unless abslutely necessary
 - [ ] Rebase your commit history so that each commit is meaningful and states the intention behind the change
 - [ ] Ensure that your pull request follows the [style guidelines](https://github.com/AutolabJS/AutolabJS/wiki/Coding-Standards)
 - [ ] At least one commit should mention `Fixes #<issue number>` _if relevant_

## Notify reviewers
@prasadtalasila

The sequence of installation commands are:
```shell
> #clone the repository
> mkdir autolabjs-install
> cp -rf AutolabJS/* autolabjs-install/.
> cd autolabjs-install/deploy
> bash dependencies.sh single_machine
> # configure AutolabJS and load docker images
> sudo ansible-playbook playbook.yml
```

### CODE REVIEW

File: deploy/setup.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+- hosts: machines
+  become: yes
+  vars:
+    main_server_public: '{{ install_dir }}/main_server/public'
+  tasks:
+
+    - name: Installing npm in main_server
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Install node packages mentioned in package.json for main server

File: deploy/setup.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      npm:
+        path: '{{ install_dir }}/main_server'
+
+    - name: Installing npm in main_server/public/js
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Install node packages mentioned in package.json for user-facing website of AutolabJS

File: deploy/setup.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      npm:
+        path: '{{ main_server_public }}/js'
+
+    - copy:
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Copy the required website dependencies to correct location

File: deploy/setup.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        src: '{{ main_server_public }}/js/node_modules/file-saver/FileSaver.min.js'
+        dst: '{{ main_server_public }}/js'
+    - copy:
+        src: '{{ main_server_public }}/js/node_modules/materialize-css/dist/js/materialize.min.js'
+        dst: '{{ main_server_public }}/js'
+    - copy:
+        src: '{{ main_server_public }}/js/node_modules/materialize-css/dist/css/materialize.min.css'
+        dst: '{{ main_server_public }}/css'
+
+    - name: Removing the node_modules directory
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Remove node_modules directory of AutolabJS website

File: deploy/setup.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    - file:
+        state: absent
+        path: '{{ main_server_public }}/js/node_modules/'
+
+    - name: Installing npm in load_balancer
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Install node packages mentioned in package.json for load balancer

File: deploy/setup.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      npm:
+        path: '{{ install_dir }}/load_balancer' 
+
+    - name: Installing npm in execution nodes
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Install node packages mentioned in package.json for execution nodes

File: deploy/setup.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      npm:
+        path: '{{ install_dir }}/execution_nodes/execution_node_{{ item }}'
+      with_sequence: start=1 count={{ number_of_execution_nodes }}
+
+    - name: Installing npm in util directory
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Install node packages mentioned in package.json for util/ module

File: deploy/setup.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      npm:
+        path: '{{ install_dir }}/util'
+
+    - debug:
+        msg: "Creating SSL Certificates"
+
+    - name: Running keys.sh
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Generate SSL keys using keys.sh

File: deploy/setup.yml [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      shell: '{{ playbook_dir }}/keys.sh'
```
Comment (prasadtalasila) [inline deploy/setup.yml]: Once the playbook is moved to plays/, the path shall be 'cd {{ playbook_dir }}/.. && bash keys.sh'

File: deploy/setup.yml [hunk 10]
```diff
@@ -0,0 +1,55 @@
+- hosts: machines
```
Comment (prasadtalasila) [inline deploy/setup.yml]: The setup.yml file should be placed in `deploy/plays/` directory.

**Comment by prasadtalasila** [comment]
@shivin7 Please rebase your branch with the latest commits on the dev branch. You seem to be directly editing code in the GitHub website. Please make use of a code editor (IDE) and push slightly bigger commits. The existing commits can be squashed to one bigger commit.

**Comment by shivin7** [comment]
Sir, I've made the requested changes, as well as squashed my commits for the creation of `setup.yml` into a single one.

File: deploy/setup.sh [hunk 11]
```diff
@@ -1,45 +1,11 @@
 #!/bin/bash
```
Comment (prasadtalasila) [inline deploy/setup.sh]: You can remove this file completely and add the line
```
- import_playbook: plays/setup.yml
```
between lines 2 and 3 of playbook.yml file.

File: deploy/plays/setup.yml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+    - name: Install node packages mentioned in package.json for util/ module
+      npm:
+        path: '{{ install_dir }}/util'
+
+    - debug:
+        msg: "Creating SSL Certificates"
+
+    - name: Generate SSL keys using keys.sh
+      shell: '{{ install_dir }}/deploy/keys.sh'
```
Comment (prasadtalasila) [inline deploy/plays/setup.yml]: This command needs to be executed from `deploy/` directory. Hence the correct Ansible task shall be as follows.
```
    - name: Generate SSL keys using keys.sh
      shell: keys.sh
      args:
        chdir: '{{ install_dir }}/deploy/'
        executable: /bin/bash
```

Ref: Ansible [shell module docs](http://docs.ansible.com/ansible/latest/modules/shell_module.html)

File: deploy/plays/setup.yml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    - debug:
```
Comment (prasadtalasila) [inline deploy/plays/setup.yml]: Since we are adding this play to the playbook.yml, the debug message can be removed.

**Comment by prasadtalasila** [comment]
@shivin7 Since we are completely deleting `setup.sh`, there is no further need for `setup.conf`. Both files can be removed.

**Comment by shivin7** [comment]
Sir, I've made the requested changes.

**Comment by prasadtalasila** [comment]
@shivin7 The changes look good. Please go ahead and add the `copy.yml` play to this PR as well. Of course, you need to add the `copy.yml` play to the `playbook.yml` before `setup.yml` play. Thanks.

**Comment by prasadtalasila** [comment]
@shivin7 One more suggestion: Please replace line-30 of [waiting task](https://github.com/AutolabJS/AutolabJS/blob/dev/deploy/plays/gitlab.yml#L30) in `gitlab.yml` with the following line.
```
   shell: docker exec gitlab bash -c "/wait-for-it.sh -t 600 127.0.0.1:8080"
```

**Comment by shivin7** [comment]
Sir, I've created the `copy.yml` play replacing the `copy.sh`. Also, I've made the change to `gitlab.yml` as requested.
Kindly take a look. I'm sorry for the delay.

File: deploy/plays/copy.yml [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    - file:
+        path: '{{ install_dir }}/log/load_balancer'
+        state: directory
+    - copy:
+        src: 'load_balancer/'
+        dst: '{{ install_dir }}/load_balancer/'
+      args:
+        chdir: '{{ playbook_dir }}/../'
+
+    - name: Creating directories for execution nodes
```
Comment (prasadtalasila) [inline deploy/plays/copy.yml]: As per Ansible docs of [copy module](http://docs.ansible.com/ansible/latest/modules/copy_module.html), the destination directory is created if it does not exist. Please see the explanation under `dest` property of the page. Hence, we don't need to create the following directories:
```
{{ install_dir }}/execution_nodes/execution_node_{{ item }}
```

**Comment by shivin7** [comment]
Sir, I've made the requested changes. Kindly check.
Also, there was a typo in `setup.yml` and `copy.yml` wherein `dest` attribute of `copy` task had been written as `dst`. I have corrected that as well in the latest commit.

**Comment by prasadtalasila** [comment]
@shivin7 Overall changes look good. Good work.

@AnkshitJain Please run the installation once on the deployment server and check. The sequence of installation commands are:
```
> #clone the repository
> mkdir autolabjs-install
> cp -rf AutolabJS/* autolabjs-install/.
> cd autolabjs-install/deploy
> # configure and load docker images
> sudo ansible-playbook playbook.yml
```

**Comment by prasadtalasila** [comment]
@AnkshitJain Please update the deployment status of this PR. This PR seems close to finish line. It should be merged into the dev branch if it works.

I have given some thought to the installation procedure. Given how difficult it is to get the dependencies installed and docker built, it is better to have two installation commands.
```shell
> bash script/deps 
> #perform the necessary configuration
> bash script/install
```
The deps script can change to deploy directory, perform ansible installation, run playbooks for downloading docker images, build the necessary docker images, copy the files to the selected installation directory.

After that we can perform the necessary configuration of the SSL certificates, AutolabJS components. We follow it up with one shell script which has the playbook invocation. This process is similar to the classic UNIX installation steps of `configure`, `make`, `make install`. We are swapping the first two steps to make the installation as `make`, `configure`, `make install`. For microservices and distributed software, this altered sequence makes sense.

@shivin7 You were working on keys.sh. Have you made enough progress on converting keys.sh to a play to be added to this PR? If not, we can consider that work to be part of another PR.

Thanks guys.

**Comment by prasadtalasila** [comment]
PR #301 improves upon the changes proposed here. Hence closing this PR.

**Pull request closed without merging**